### ISSUE TITLE
mark_safe problem

### ISSUE THREAD

**Issue by nbob** [asyncee/django-easy-select2#44]
How can I write a custom html to output options?

**Comment by asyncee** [comment]
Hey, you can override rendering method:

https://github.com/asyncee/django-easy-select2/blob/master/easy_select2/widgets.py#L95

https://github.com/asyncee/django-easy-select2/blob/771e2924fb39578630c9a9cabacb34d5e66b2cdf/easy_select2/widgets.py#L95
https://github.com/asyncee/django-easy-select2/blob/771e2924fb39578630c9a9cabacb34d5e66b2cdf/easy_select2/widgets.py#L95
https://github.com/asyncee/django-easy-select2/blob/771e2924fb39578630c9a9cabacb34d5e66b2cdf/easy_select2/widgets.py#L95

**Comment by nbob** [comment]
I advice to rewrite whole function and replace only one line?
It doesn't look like DRY solution

May it would be better to allow developers to decide by them self escape or not escape?

**Comment by asyncee** [comment]
@nbob Sure, it would be better, what is your use case exactly?

PR's are welcomed also.

@gauravv7 what is your thoughts on this?

**Comment by nbob** [comment]
> Sure, it would be better, what is your use case exactly?
Use case is very simple. I have a list of choices with icons and title of them. So, I just want to display select2 widget with html-options in my django admin

> PR's are welcomed also.
I see that `mark_safe` is present in `render_method`, so, I have no idea whats the problem in

**Comment by gauravv7** [comment]
I agree with @asyncee, overriding render function gives a lot of flexibility which cannot be generalized e.g. Nested html, event bubbling/capturing, analytics handlers etc. 
But for a simple requirement like nbob's, a provision can be made. I think we can do it for making the plugin more flexble to use. I am preparing a PR for this, @asyncee let me know what you think of this

**Comment by nbob** [comment]
@asyncee @gauravv7 
I've read a bit of select2 docs and find out that `select2` escapes option's text by default.
But `select2` allows to override escape function by setting `escapeMarkup` option in select, but there is no way to pass it through `django-easy-select2` lib

**Comment by gauravv7** [comment]
@nbob as @asyncee suggested, you need to override the rendering method. Take a look at https://github.com/asyncee/django-easy-select2/blob/master/easy_select2/widgets.py#L76
Which is for rending select2 options code.

Also please take a look at https://github.com/asyncee/django-easy-select2/blob/master/easy_select2/widgets.py#L71
For all the options specified to select2 options(js plugin) to its constructor.

**Comment by asyncee** [comment]
@gauravv7 you are absolutely correct. I would be great if you can do a PR.

As for nbob's request — can you show how final (rendered) select2 template should be to better understand the problem?

**Comment by nbob** [comment]
@gauravv7 
> For all the options specified to select2 options(js plugin) to its constructor.
Yes, I see.
1. I need to pass escapeHtml function to select2
2. It is not possible to pass it via to data-* attribute, because select2 does not use eval on its data-* attribute string, fortunately.
3. The only one way is to pass it as pure JS function in $.select2 constructor func. But this is not allowed by `django-easy-select2`
4. Overriding `render` method does not give any result, because it does not control JS code in `/js/init.js` file

@asyncee 
> can you show how final (rendered) select2 template should be to better understand the problem?
It doesn't matter, but for example I want smth like this in my option tag
`<div><i class='ai ai-some-icon-cls'></i> Some icon title</div>`

**Comment by asyncee** [comment]
I think we can extend plugin to allow custom select2 initialization with following (little bit hacky it looks for me anyway):

```
// define global select2 configuration callback
// this callback will be called in redisplay_select2()
function initSelect2Custom(domNode) {
    // custom initialization code here
} 
```

To let select2 initialization code know that it need to use custom callback, custom data attribute may be passed (data-custom-callback="initSelect2Custom")

On initialization, function `add_select2_handlers()` can determine if custom callback name is provided and use it via `window[$(domNode).data('custom-callback')]`.

So, to provide custom initialization function one should be doing something like this:

1. define custom initialization callback on `window`
2. tell select2 widget to use it via Select2(..., init_callback='initSelect2Custom')

So what do you think? Are there easier way?

**Comment by gauravv7** [comment]
This issue is resolved with a new release of version 1.5.0. Ergo, Closing this issue for now.
